robot/lib/Src/drivers/ICM20948.cpp

```diff
@@ -0,0 +1,180 @@
+#include "drivers/ICM20948.hpp"
+#include <cmath>
+
+#define READ_BIT (1 << 7)
+
+#define WHOAMI_VAL 0xEA
+
+namespace Registers {
+    // Bank select
+    constexpr uint8_t REG_BANK_SEL  = 0x7F;
+
+    // Bank 0
+    constexpr uint8_t WHO_AM_I      = 0x00;
+    constexpr uint8_t USER_CTRL     = 0x03;
+    constexpr uint8_t LP_CONFIG     = 0x05; 
+    constexpr uint8_t PWR_MGMT_1    = 0x06;
+    constexpr uint8_t PWR_MGMT_2    = 0x07;
+
+    constexpr uint8_t GYRO_Z_OUT_H  = 0x37;
+    constexpr uint8_t GYRO_Z_OUT_L  = 0x38;
+
+    constexpr uint8_t FIFO_MODE     = 0x69;
+   
+    // Bank 2 
+    constexpr uint8_t GYRO_CONFIG_1 = 0x01;
+
+    enum GYRO_FS_SEL {
+        DPS_2000 = 3,
+        DPS_1000 = 2,
+        DPS_500 = 1,
+        DPS_250 = 0,
+    };
+
+    enum PWR_MGMT {
+        DEVICE_RESET = 1,
+        SLEEP = 1,  // 1: enter sleep mode
+        CLKSEL = 1, // 1-5: Auto selects best available clock source
+    };
+
+    enum LP_CONFIG_REG {
+        I2C_MST_CYCLE = 1,
+    };
+
+    enum USER_CTRL_REG {
+        I2C_IF_DIS = 1, // disable I2C, enable SPI
+    }; 
+    
+};
+
+ICM20948::ICM20948(std::shared_ptr<SPI> imuSPI, PinName cs_pin) 
+    : imuSPI(std::move(imuSPI)),
+      nCs(cs_pin) {
+    //chip_select_pin.write(false);
+    nCs = 1;
+    HAL_Delay(1500);
+}
+
+void ICM20948::initialize() {
+    // SPI pulse to clear pins. 
+    imuSPI->transmit(0);    
+    
+    uint8_t whoami = read_register(0, Registers::WHO_AM_I);
+    
+    while (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        whoami = read_register(Registers::WHO_AM_I);
+    }
+    
+    // Reset
+    write_register( 
+        Registers::PWR_MGMT_1,
+        Registers::PWR_MGMT::DEVICE_RESET << 7
+        | Registers::PWR_MGMT::SLEEP << 6
+        | Registers::PWR_MGMT::CLKSEL);
+
+    HAL_Delay(100);
+
+    // Check whoami again
+    whoami = read_register(0, Registers::WHO_AM_I);
+    
+    while (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        whoami = read_register(Registers::WHO_AM_I);
+    }
```

Comment:
Yeah, this shouldn't really be a "failure" so much as a "waiting for IMU to reboot".

It should also very much not be a spin loop, because the IMU runs at a super high priority and this will choke out other threads. Just add a `vTaskDelay` in.